### PR TITLE
Bug fix: feature/issue-7804-chroma-client_settings-bug

### PR DESCRIPTION
**Pull request by kristellewidjaja**
Description: Made Chroma constructor more robust when client_settings is provided. Otherwise, existing embeddings will not be loaded correctly from Chroma.
Issue: #7804
Dependencies: None
Tag maintainer: @rlancemartin, @eyurtsev

### CODE REVIEW

File: libs/langchain/langchain/vectorstores/chroma.py [hunk 1]
```diff
@@ -92,6 +92,13 @@ def __init__(
             self._persist_directory = persist_directory
         else:
             if client_settings:
+                # Check the client settings required input values
+                if client_settings.chroma_db_impl is None or \
+                    client_settings.chroma_db_impl == "":
+                    raise ValueError("client_settings.chroma_db_impl is required")
+                if client_settings.persist_directory is None or \
+                    client_settings.persist_directory == "":
+                    raise ValueError("client_settings.persist_directory is required")
```
Comment (baskaryan) [inline libs/langchain/langchain/vectorstores/chroma.py:101]: is this actually required?

**Review by baskaryan — COMMENTED** [review]
cc @jeffchuber

File: libs/langchain/langchain/vectorstores/chroma.py (hunk 1 above)
Comment (jeffchuber) [reply libs/langchain/langchain/vectorstores/chroma.py:101]: No, it's not. Chroma has 3 modes

`chromadb.EphemeralClient()` - runs in memory
`chromadb.PersistentClient()` - runs in memory and saves to disk
`chromadb.HttpClient()` - talks to a remote chroma server

**Comment by jeffchuber** [comment]
@kristellewidjaja this commit also seems wrong... in chroma `0.4` - we no longer support chroma_db_impl - https://github.com/chroma-core/chroma/blob/main/chromadb/__init__.py. What version of Chroma are you using?

**Comment by kristellewidjaja** [comment]
@jeffchuber Thank you for providing your insights! FYI, I am using the chromadb.PersistentClient() mode - runs in memory and saves to disk. 

I was using langchain==0.0.240 and chromadb==0.3.26. 

Just to be clear, the issue seems to only occur in the Langchain Chroma() constructor when client_settings is explicitly provided:

1. When client_settings.persist_directory is provided and chroma_db_impl is null, my existing embeddings are not loaded. I propose that chroma_db_impl should be required in this case. 
2. When client_settings.persist_directory is null and chroma_db_impl is provided, it raises a different error: “ValueError: You cannot use chroma's cache directory .chroma/, please set a different directory”
3. When client_settings.persist_directory is null and chroma_db_impl is also null, my existing embeddings are not loaded.

If you need more information, please refer to the [original issue](https://github.com/langchain-ai/langchain/issues/7804) logged by another user. 

I also tried upgrading to chromadb==0.4.3 and passed in the client object to Langchain Chroma() constructor but my existing embeddings were still not loaded.  

I have updated my PR to only enforce the necessary input parameter to the Langchain Chroma() constructor when running in-memory and persisting to disk mode. Please let me know what you think.

**Comment by jeffchuber** [comment]
@kristellewidjaja ahhh I see the issue. This is a cross version issue where `0.4` drops user-specified `chroma_db_impl` ... can we use a version check like this? https://github.com/langchain-ai/langchain/blob/23f953e8ab81e086c4e1de5f4d200012d767b7b9/libs/langchain/langchain/vectorstores/chroma.py#L106

thanks for catching this!

**Comment by kristellewidjaja** [comment]
@jeffchuber Thanks! I applied your suggestion to my PR. Now, it works for the cases. Please verify it when you get a chance. 

Screenshot 2023-08-01 at 12 21 43 AM

**Comment by jeffchuber** [comment]
@kristellewidjaja works for me 🎉

**Comment by kristellewidjaja** [comment]
@jeffchuber thanks for the quick turnaround! i've been using ChromaDB and I love it!

@baskaryan what is the next step to have my PR reviewed and possibly merged to the main branch?

**Comment by peidaqi** [comment]
I think this PR #8267 introduced a new issue where is_persistent also needs to be set to True for persistent directory to work (ChromaDB defaults is_persistent to False)

See this PR for details:

https://github.com/langchain-ai/langchain/pull/8733

**Comment by jeffchuber** [comment]
This init logic has become extremely confusing (and therefore hard to maintain) - I think I need to rewrite it entirely.

**Comment by peidaqi** [comment]
@jeffchuber agreed. Maybe at certain point lock ChromaDB dependency to a newer version and drop duckdb+parquet support completely.

**Comment by jeffchuber** [comment]
@peidaqi haha - we did try to do that! we built a CLI around this to help users, https://github.com/chroma-core/chroma-migrate

but the consensus was from others that backwards compatibility was ideal.

I'll take a pass tonight

**Pull request merged**